azurerm/internal/services/sentinel/sentinel_data_connector_office_365.go

```diff
@@ -0,0 +1,244 @@
+	if !exchangeEnabled && !sharePointEnabled && !teamsEnabled {
+		return fmt.Errorf("one of `exchange_enabled`, `share_point_enabled` and `teams_enabled` should be `true`")
+	}
+
+	exchangeState := securityinsight.Enabled
+	if !exchangeEnabled {
+		exchangeState = securityinsight.Disabled
+	}
+
+	sharePointState := securityinsight.Enabled
+	if !sharePointEnabled {
+		sharePointState = securityinsight.Disabled
+	}
+
+	teamsState := securityinsight.Enabled
+	if !teamsEnabled {
+		teamsState = securityinsight.Disabled
+	}
+
+	param := securityinsight.OfficeDataConnector{
+		Name: &name,
+		OfficeDataConnectorProperties: &securityinsight.OfficeDataConnectorProperties{
+			TenantID: &tenantId,
+			DataTypes: &securityinsight.OfficeDataConnectorDataTypes{
+				Exchange: &securityinsight.OfficeDataConnectorDataTypesExchange{
+					State: exchangeState,
+				},
+				SharePoint: &securityinsight.OfficeDataConnectorDataTypesSharePoint{
+					State: sharePointState,
+				},
+				Teams: &securityinsight.OfficeDataConnectorDataTypesTeams{
+					State: teamsState,
+				},
+			},
+		},
+		Kind: securityinsight.KindOffice365,
+	}
+
+	// Service avoid concurrent updates of this resource via checking the "etag" to guarantee it is the same value as last Read.
+	if !d.IsNewResource() {
+		resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, name)
+		if err != nil {
+			return fmt.Errorf("retrieving Sentinel Data Connector Office 365 %q: %+v", id, err)
+		}
+
+		if err := assertDataConnectorKind(resp.Value, securityinsight.DataConnectorKindOffice365); err != nil {
+			return fmt.Errorf("asserting Sentinel Data Connector of %q: %+v", id, err)
+		}
+		param.Etag = resp.Value.(securityinsight.OfficeDataConnector).Etag
+	}
+
+	_, err = client.CreateOrUpdate(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name, param)
+	if err != nil {
+		return fmt.Errorf("creating Sentinel Data Connector Office 365 %q: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceSentinelDataConnectorOffice365Read(d, meta)
+}
+
+func resourceSentinelDataConnectorOffice365Read(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Sentinel.DataConnectorsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.DataConnectorID(d.Id())
+	if err != nil {
+		return err
+	}
+	workspaceId := loganalyticsParse.NewLogAnalyticsWorkspaceID(id.SubscriptionId, id.ResourceGroup, id.WorkspaceName)
+
+	resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Sentinel Data Connector Office 365 %q was not found - removing from state!", id)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving Sentinel Data Connector Office 365 %q: %+v", id, err)
```

Comment:
we can reduce this to:

```suggestion
		return fmt.Errorf("retrieving %s: %+v", id, err)
```